### DISCUSSION TITLE
Do other graphql framework need to add support for Router context manipulate?

### DISCUSSION THREAD

**Discussion by BCantos17** [apollographql/router#929] [Q&A] [answered]
I am experimenting with router and I am running this app (preview 6)

https://github.com/apollographql/router/tree/main/examples/context

I did not change anything and I am inserting the supergraph schema I have which was generated by apollo studios. The subgraphs are using the Netflix DGS framework. After I start up the router and subgraphs. I try viewing what's inside the context in the subgraph but I do not see the value `"incoming_data": "world!"` anywhere. I got this working in apollo gateway so this is making me wondering if DGS needs to somehow enable support for it. Can someone confirm this?

**Comment by bnjjj** [comment, marked as answer]
You don't need to add support for the router in subgraphs it should be supported by default. Are you able to see the response of your query ? Did you launch the router with the same yaml config than in the example ?

**Reply by BCantos17** [reply]
Yup I can see the response of my query. I didn't change a thing on the yaml config, it's untouched

**Reply by bnjjj** [reply]
Ok after testing the `context` example I can see the logs with the data inside context. So after re-reading your initial message I discovered that I maybe misunderstood your question/needs. You're saying `inside the context in the subgraph` but just to clarify, the context we have in the router only exists in the router and during the lifecycle of a request. Could you elaborate about what you're trying to do ? What do you expect from context ? You can find more documentation about context in [our docs](https://www.apollographql.com/docs/router/customizations/native/#5-define-necessary-context)

**Reply by BCantos17** [reply]
Sure. Basically I want to write a plugin to attach additional information in the context before the initial request to the subgraph, I need to do this for authentication reasons

The flow will be like this:
1. user makes a request
2. router takes the request
3. router make call needed for authentication before the call to the subgraph
4. router attaches data for authentication to it's context
5. router will send both the context and the request to the subgraph
6. subgraph reads the data in context for authentication and completes request

This flow worked in apollo gateway but I am either failing to pass data in the context in the router or failing to read the data in the subgraph. I am using the DGS framework for my subgraphs and I know they haven't supported federation 2 yet so this makes me think that it's possible router is doing something different when passing data in the context and DGS just doesn't support the way router does things yet. Or I could just be messing things up.

**Reply by bnjjj** [reply]
OK then if you want to pass some data from the router to a subgraph you should pass it in the headers. To do your flow, if it's the same authentication data for each subgraphs then you can fetch your authentication data from the `router_service` for example, put the authentication data into the context (like the auth token or something else). Then in `subgraph_service` you will be able to use something like 

```rust
service.map_request(|mut req| {
    let my_token: String = req.context.get("auth_data").unwrap().unwrap();
    // auth_data is the key you previously set when inserting auth data in the context in router_service
    req.subgraph_request.headers_mut().insert("my_header_name", my_token);
    req
}).boxed() 
```

**Reply by BCantos17** [reply]
That worked! Thank you so much @bnjjj!
